### ISSUE TITLE
Over 300,000 empty playlists created by LibreTime

### ISSUE THREAD

**Issue by smilingpeanut** [LibreTime/libretime#411]
This has happened twice now on our installation. Log in to LT to make a change on the calendar. The calendar takes several minutes to load. Look in Playlists and there are over 350,000 empty playlists that have been created by the system. The solution is to manually go into the database and use an SQL command (`DELETE FROM cc_playlist WHERE creator_id IS NULL;`) to clear out the playlists.

Does anyone know why this happens?

<img width="890" alt="screen shot 2018-01-17 at 11 18 17 pm" src="https://user-images.githubusercontent.com/1864093/35083795-1be2521e-fbde-11e7-965a-5706216d27ce.png">

**Comment by Robbt** [comment]
This indeed is quite odd. The only mechanism I can think of that would be making playlists off the top of my head is the Podcast option to create a playlist when a new podcast is added. Is this enabled under settings ? Just trying to determine if possibly it is some unintended side effect of that change. Otherwise I'm not sure where to start with this one as thats the only code I touched that created playlists. If that was the case it'd also create smartblocks and they also shouldn't be blank. So this doesn't really explain what is happening here. Do you have any sort of cronjob or anything running ?

**Comment by smilingpeanut** [comment]
Yeah, the setting to create playlists for a podcast is enabled. I can try disabling it. There are smartblocks for each podcast, but they're the correct number. The only anomaly is the number of playlists. No cronjob running.

**Comment by Robbt** [comment]
Let us know if it keeps occurring and if there is any event that appears to be triggering it. I'm going to look into the playlist code related to another bug. Perhaps a recent PR is having unintended consequences.

**Comment by smilingpeanut** [comment]
It occurred again, so I disabled the setting to automatically create playlists/smartblocks for each new podcast. A few days later, after I'd cleared out all the playlists again, there were about 150 new blank playlists.

**Comment by Robbt** [comment]
Hmm, so this is odd. Anything unique about your LibreTime setup ?
Anything in your zend-php.log under /etc/airtime to indicate this happening or perhaps your airtime-celery.log

**Comment by smilingpeanut** [comment]
Nothing unique that I can think of. No customizations or deviations from the norm. I searched the server for both of those log files and they don't exist.

**Comment by Robbt** [comment]
Sorry it should be under /var/log/airtime I believe. What files are in there.

**Comment by smilingpeanut** [comment]
Okay, so I've got `airtime_analyzer.log` and `zendphp.log` in that directory. Which portions of these files might be relevant? I'd rather not post them publicly.

**Comment by Robbt** [comment]
I would just search thru them and try to find any errors or something repeating enough to indicate it was creating the playlists. I'm not sure what would be causing the issue you are experiencing so I'm not sure what too look for. If you do end up posting a portion of it, just hide the API key and/or host name and anything else you want to avoid making public.

**Comment by smilingpeanut** [comment]
354,000 playlists created again. Here's the `zendphp.log` file:

https://gist.github.com/smilingpeanut/728f984c63d045a6d06343b53205b82a

**Comment by Robbt** [comment]
Ok, so perhaps the issue is occurring here.
`2018-01-27T05:34:55+00:00 WARN (4): [hostname] Celery task DbId: 556
DbTaskId: php_5a6c0f10574437.89778857
DbTrackReference: 556
DbName: podcast-download
DbDispatchTime: '2018-01-27 05:33:04'
DbStatus: SUCCESS
ThirdPartyTrackReferences:
  DbId: 556
  DbService: podcast
  DbForeignId: '566'
  DbFileId: null
  DbUploadTime: '2018-01-27 05:34:55'
  DbStatus: null
  CeleryTaskss:
    CeleryTasks_0: '*RECURSION*'
episode 566 unsuccessful with message can't sync to an MPEG frame`

Not sure what is causing this - although it is something to do with Python's analysis of the file headers with mutagen - https://stackoverflow.com/questions/22093456/python-mutagen-cant-get-artist-from-mp4-mp3-file#22093541

So it is possible that it tries to read the Album from a podcast and fails which spurs up a bunch of attempts to create playlists that match the podcast name which is blank ? Still not sure why this would create thousands of playlists.

Obviously there is a bug somewhere in the podcast-download code, the RECURSION makes me think that perhaps it is spawning thousands of subprocesses creating blank playlists. 

**Did you turn off the option to enable auto-create playlists when you add podcasts or is that still enabled ?**

Are there any podcasts that have odd files ? Can you figure out what podcast is causing this issue and if possible share it so that I can try to replicate the issue. This maybe unrelated to the "unnamed playlists" - I'm assuming the names are  blank and also empty,


It is also possible that this error could be related to it although I don't see why there is an issue with scheduling or why this would create blank playlists.

this all triggered by wrapped: SQLSTATE[40001]: Serialization failure: 7 ERROR: could not serialize access due to concurrent update]

which I don't understand why scheduling would have a concurrent update 

`
2018-01-27T20:14:52+00:00 INFO (6): [hostname] Checking autoplaylist poll
2018-01-27T20:14:52+00:00 INFO (6): [hostname] Checking to run Auto Playlist
2018-01-27T20:14:52+00:00 INFO (6): [hostname] Scheduling 
2018-01-27T20:14:52+00:00 ERR (3): [hostname] [ErrorController.php:26 - errorAction()] - Unable to execute UPDATE statement [UPDATE cc_show_instances SET last_scheduled=:p1 WHERE cc_show_instances.id IN (:p2,:p3,:p4,:p5,:p6,:p7)] [wrapped: SQLSTATE[40001]: Serialization failure: 7 ERROR:  could not serialize access due to concurrent update]
2018-01-27T20:14:52+00:00 ERR (3): [hostname] [ErrorController.php:27 - errorAction()] - #0 /usr/share/airtime/php/vendor/propel/propel1/runtime/lib/query/ModelCriteria.php(1837): BasePeer::doUpdate(Object(CcShowInstancesQuery), Object(Criteria), Object(PropelPDO))
#1 /usr/share/airtime/php/vendor/propel/propel1/runtime/lib/query/ModelCriteria.php(1793): ModelCriteria->doUpdate(Array, Object(PropelPDO), false)
#2 /usr/share/airtime/php/airtime_mvc/application/models/Scheduler.php(1013): ModelCriteria->update(Array, Object(PropelPDO))
#3 /usr/share/airtime/php/airtime_mvc/application/models/Scheduler.php(1080): Application_Model_Scheduler->insertAfter(Array, Array, NULL, true)
#4 /usr/share/airtime/php/airtime_mvc/application/models/ShowInstance.php(245): Application_Model_Scheduler->scheduleAfter(Array, Array)
#5 /usr/share/airtime/php/airtime_mvc/application/common/AutoPlaylistManager.php(41): Application_Model_ShowInstance->addPlaylistToShow(NULL, false)
#6 /usr/share/airtime/php/airtime_mvc/application/common/TaskManager.php(248): AutoPlaylistManager::buildAutoPlaylist()
#7 /usr/share/airtime/php/airtime_mvc/application/common/TaskManager.php(61): AutoPlaylistTask->run()
#8 /usr/share/airtime/php/airtime_mvc/application/common/TaskManager.php(104): TaskManager->runTask('AutoPlaylistTas...')
#9 /usr/share/airtime/php/airtime_mvc/application/controllers/plugins/PageLayoutInitPlugin.php(72): TaskManager->runTasks()
#10 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Controller/Plugin/Broker.php(260): PageLayoutInitPlugin->routeShutdown(Object(Zend_Controller_Request_Http))
#11 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Controller/Front.php(923): Zend_Controller_Plugin_Broker->routeShutdown(Object(Zend_Controller_Request_Http))
#12 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#13 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#14 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(84): Zend_Application->run()
#15 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')`



Also do you have a airtime-celery.log file ? 

That might have more details about what is happening behind the scenes with the podcast downloads and what is going awry. I'm suspecting the error is with my code adding smartblocks and playlists for new podcasts going haywire when there is something improperly parsed about the podcast. 

Have you added any new podcasts around the time this error has occurred ? Or does it just happen out of the blue.

**Comment by guythnick** [comment]
I have had the same issue happen to me twice as well.  At 5:00AM this morning Libretime created about 285k blank playlists.  I have never had the Podcast Automatic Smartblock and Playlist setting enabled, so I don't think that is it.  I have been running this to fix it, as it slows down the site quite a bit:
` delete  from cc_playlist where name = '' and mtime is null;`

**Comment by Robbt** [comment]
Can you check your zend-php.log file under /var/log/airtime and also your airtime-celery.log and see if they contain errors. It would be helpful to see if any of the errors or log messages correspond with the "Uploaded" column for the playlists so we can see if the creation of the empty playlists corresponds to a specific error.

@guythnick Are you using the podcast functionality ?

**Comment by guythnick** [comment]
I am not using podcast functionality at all.  Also, I do not see any celery log files.  Are they in a different location?  Below is the zend log, there are a ton of errors..: 

[zendphp.log](https://gist.github.com/guythnick/5ee9ec37b4b6497d8d2f7bcb070b6f4e)

**Comment by Robbt** [comment]
Ok, I guess since you deleted the playlists we can't see if any correspond with the errors. For some reason logs for airtime-celery aren't being created. I opened issue #442 to track that as getting the logs for celery will help us determine if it is related to that. Since you aren't using the podcast functionality at all it is unlikely that this is connected.

Where you using LibreTime when this error occurred or did it happen when you weren't logged in.
Also what distribution/version of linux did you install this on.

Thanks. Really want to figure out what is causing this as obviously there is a bug somewhere.

**Comment by guythnick** [comment]
Firstly, thanks a ton for maintaining this and looking into the issue. Hopefully I can help out in the future once I wrap my head around the code / database schema. 
There could have been some live streaming during the time the playlists were created, but I'm not for sure.  I will say regarding the playlists that we only have 3 currently, with IDs 1,2,3.  So anything that would reference a playlist that wasn't one of those would be one of the blank ones. 

I am running Libretime 0327a75 on Ubuntu Server 14.04.5 LTS, which is a VM guest of ESXi 6.5.

**Comment by guythnick** [comment]
This happened again on the 17th.  Started at 5:00am also.  I think it has something to do with the fact I have 3 smart blocks that I also have each one in 3 separate playlists.  These playlists are used to auto-schedule some of the shows.  What I don't get is why at 5:00am this happens sometime.  Some sort of weekly cron job?

date_trunc('hour',utime), count(*)
"2018-02-17 07:00:00"	"14962"
"2018-02-17 06:00:00"	"243499"
"2018-02-17 05:00:00"	"45"

**Comment by Robbt** [comment]
So the error consistent between both of these bug reports is
```
2018-02-09T23:01:48-06:00 ERR (3): [hostname] [ErrorController.php:26 - errorAction()] - Unable to execute UPDATE statement [UPDATE cc_schedule SET playout_status=:p1 WHERE cc_schedule.instance_id=:p2 AND cc_schedule.playout_status>=:p3 AND cc_schedule.ends<=:p4] [wrapped: SQLSTATE[40001]: Serialization failure: 7 ERROR:  could not serialize access due to concurrent update]
2018-02-09T23:01:48-06:00 ERR (3): [hostname] [ErrorController.php:27 - errorAction()] - #0 /usr/share/airtime/php/vendor/propel/propel1/runtime/lib/query/ModelCriteria.php(1837): BasePeer::doUpdate(Object(CcScheduleQuery), Object(Criteria), Object(PropelPDO))
#1 /usr/share/airtime/php/vendor/propel/propel1/runtime/lib/query/ModelCriteria.php(1793): ModelCriteria->doUpdate(Array, Object(PropelPDO), false)
#2 /usr/share/airtime/php/airtime_mvc/application/models/airtime/CcShowInstances.php(121): ModelCriteria->update(Array, Object(PropelPDO))
#3 /usr/share/airtime/php/airtime_mvc/application/models/Scheduler.php(1001): CcShowInstances->updateScheduleStatus(Object(PropelPDO))
#4 /usr/share/airtime/php/airtime_mvc/application/models/Scheduler.php(1080): Application_Model_Scheduler->insertAfter(Array, Array, NULL, true)
#5 /usr/share/airtime/php/airtime_mvc/application/models/ShowInstance.php(246): Application_Model_Scheduler->scheduleAfter(Array, Array)
#6 /usr/share/airtime/php/airtime_mvc/application/common/AutoPlaylistManager.php(41): Application_Model_ShowInstance->addPlaylistToShow(NULL, false)
#7 /usr/share/airtime/php/airtime_mvc/application/common/TaskManager.php(248): AutoPlaylistManager::buildAutoPlaylist()
#8 /usr/share/airtime/php/airtime_mvc/application/common/TaskManager.php(61): AutoPlaylistTask->run()
#9 /usr/share/airtime/php/airtime_mvc/application/common/TaskManager.php(104): TaskManager->runTask('AutoPlaylistTas...')
#10 /usr/share/airtime/php/airtime_mvc/application/controllers/plugins/PageLayoutInitPlugin.php(72): TaskManager->runTasks()
#11 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Controller/Plugin/Broker.php(260): PageLayoutInitPlugin->routeShutdown(Object(Zend_Controller_Request_Http))
#12 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Controller/Front.php(923): Zend_Controller_Plugin_Broker->routeShutdown(Object(Zend_Controller_Request_Http))
#13 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#14 /usr/share/airtime/php/vendor/zendframework/zendframework1/library/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#15 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(84): Zend_Application->run()
#16 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')
```

So we still need to figure out what is causing this and how it could be creating such an absurd number of playlists. @smilingpeanut and @guythnick are either of you still running LibreTime and/or running into this problem with the latest version ?

**Comment by Robbt** [comment]
This seems like it might be the culprit - ```#6 /usr/share/airtime/php/airtime_mvc/application/common/AutoPlaylistManager.php(41): Application_Model_ShowInstance->addPlaylistToShow(NULL, false)```

I wonder if in both of these cases there was an autoloading playlist that wasn't set and then it tried to add a NULL playlist and something went widely out of control resulting in this happening thousands of times and creating blank playlists. I think I actually ran into this while working on the intro playlist as I saw a couple of totally empty playlists added seemingly out of nowhere but not the thousands mentioned in this issue.

**Comment by smilingpeanut** [comment]
@Robbt We switched from LibreTime to Radiologik DJ/Scheduler as we needed more flexibility.

**Comment by Robbt** [comment]
Quick question @arifuddin2007 - do you have autoloading playlists enabled for any of your shows.

If so do you have any of them with this enabled that have no playlist selected ? This I suspect might be the cause, and if I can confirm this is the setup on your instance I can get closer to devising a fix.

Also scan your /var/log/airtime/zend-php.log and post any errors you have in there up here (feel free to conceal any IP or host information).

**Comment by arifuddin2007** [comment]
Yes. We have configured all the shows in autoloading playlist.

I am attaching log file.

[zendphp.log.zip](https://github.com/LibreTime/libretime/files/2877023/zendphp.log.zip)

**Comment by Robbt** [comment]
For one of the shows with the autoloading playlist do you have None selected for the drop-down ? Are there other logs - I don't see any errors in there.

Also what version of LibreTime are you using ?

**Comment by Robbt** [comment]
Ok, I have figured this out. It is caused by having an Autoloading Playlist with None. I will put together a fix to avoid this behavior again.

**Comment by Robbt** [comment]
I am testing the fix I pushed through but if you want to get rid of the empty playlists you can use the following set of command on the command line.
`psql -h localhost -U airtime -W`
That will get you into the postgresql structure where you can then type
`delete from cc_playlist where name = '' and mtime is null;`

**Comment by Robbt** [comment]
So I finally figured out what caused this wicked bug and the fix is in #745 - but basically for some reason when you attempt to schedule an None aka NULL playlist through the autoloading playlist it has the side effect of creating an empty playlist (not sure why exactly). When you have repeat until fill it attempts to do this until the system starts crashing resulting in the ludicrous numbers of playlists that have been reported here. 

So this run away loop is closed with my PR and if you attempt to schedule a null playlist and a null intro playlist and null outro playlist it will do nothing and call the autoloading playlist built.
